WebCodecsVideoFrame::copyTo should support kCVPixelFormatType_Lossless_420YpCbCr8BiPlanarVideoRange

youennf opened the pull request:
#### 7a6d05086eb0690c68f3976659a12df46d3500ec
<pre>
WebCodecsVideoFrame::copyTo should support kCVPixelFormatType_Lossless_420YpCbCr8BiPlanarVideoRange
<a href="https://rdar.apple.com/151913227">rdar://151913227</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=293478">https://bugs.webkit.org/show_bug.cgi?id=293478</a>

Reviewed by Jean-Yves Avenard.

We can create a WebCodecs VideoFrame object from a video element generating kCVPixelFormatType_Lossless_420YpCbCr8BiPlanarVideoRange pixel buffers.
But we were not able to copyTo the data.

To support this, we do a conversion to NV12 in GPUProcess before sending the converted data to WebProcess.

* LayoutTests/http/wpt/webcodecs/videoFrame-video-element-compressed-expected.txt: Added.
* LayoutTests/http/wpt/webcodecs/videoFrame-video-element-compressed.html: Added.
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/cv/PixelBufferConformerCV.cpp:
(WebCore::PixelBufferConformerCV::createPixelConformer):
(WebCore::PixelBufferConformerCV::PixelBufferConformerCV):
* Source/WebCore/platform/graphics/cv/PixelBufferConformerCV.h:
* Source/WebCore/platform/graphics/cv/PixelBufferConformerCV.mm: Added.
(WebCore::PixelBufferConformerCV::PixelBufferConformerCV):
* Source/WebKit/GPUProcess/media/RemoteVideoFrameObjectHeap.cpp:
(WebKit::RemoteVideoFrameObjectHeap::convertFrameBuffer):
(WebKit::RemoteVideoFrameObjectHeap::get):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/webrtc/SharedVideoFrame.cpp:
(WebKit::SharedVideoFrameWriter::writeBuffer):
* Source/WebKit/WebProcess/GPU/webrtc/SharedVideoFrame.h:

Canonical link: <a href="https://commits.webkit.org/295421@main">https://commits.webkit.org/295421@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1686dc51570ee7a492e31580958970c418ba61a5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/105035 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24749 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/15172 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/110252 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55715 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/107076 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/25162 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/33293 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79765 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/108041 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19589 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94805 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60072 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/19344 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12881 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/55092 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/89057 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12926 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/112759 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/32200 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23704 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88845 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32566 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/91025 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88474 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22556 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33366 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/11155 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/27578 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/32124 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/37634 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31917 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/35258 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33476 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on LayoutTests/http/wpt/webcodecs/videoFrame-video-element-compressed.html:
```diff
@@ -0,0 +1,39 @@
+     if (buffer[i] != 0 && buffer[i] != 128)
+       ++isValidCount;
+  }
+  assert_greater_than(isValidCount, 160 *128);
```
```suggestion
  assert_greater_than(isValidCount, 160 * 128);
```

youennf committed:
WebCodecsVideoFrame::copyTo should support kCVPixelFormatType_Lossless_420YpCbCr8BiPlanarVideoRange
rdar://151913227
https://bugs.webkit.org/show_bug.cgi?id=293478

Reviewed by Jean-Yves Avenard.

We can create a WebCodecs VideoFrame object from a video element generating kCVPixelFormatType_Lossless_420YpCbCr8BiPlanarVideoRange pixel buffers.
But we were not able to copyTo the data.

To support this, we do a conversion to NV12 in GPUProcess before sending the converted data to WebProcess.

* LayoutTests/http/wpt/webcodecs/videoFrame-video-element-compressed-expected.txt: Added.
* LayoutTests/http/wpt/webcodecs/videoFrame-video-element-compressed.html: Added.
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/cv/PixelBufferConformerCV.cpp:
(WebCore::PixelBufferConformerCV::createPixelConformer):
(WebCore::PixelBufferConformerCV::PixelBufferConformerCV):
* Source/WebCore/platform/graphics/cv/PixelBufferConformerCV.h:
* Source/WebCore/platform/graphics/cv/PixelBufferConformerCV.mm: Added.
(WebCore::PixelBufferConformerCV::PixelBufferConformerCV):
* Source/WebKit/GPUProcess/media/RemoteVideoFrameObjectHeap.cpp:
(WebKit::RemoteVideoFrameObjectHeap::convertFrameBuffer):
(WebKit::RemoteVideoFrameObjectHeap::get):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/webrtc/SharedVideoFrame.cpp:
(WebKit::SharedVideoFrameWriter::writeBuffer):
* Source/WebKit/WebProcess/GPU/webrtc/SharedVideoFrame.h:

Canonical link: https://commits.webkit.org/295421@main
